PSReadLine/History.cs

```diff
@@ -457,12 +457,61 @@ private void ReadHistoryFile()
             {
                 WithHistoryFileMutexDo(1000, () =>
                 {
-                    var historyLines = File.ReadAllLines(Options.HistorySavePath);
+                    var historyLines = ReadHistoryLinesImpl(Options.HistorySavePath, Options.MaximumHistoryCount);
                     UpdateHistoryFromFile(historyLines, fromDifferentSession: false, fromInitialRead: true);
                     var fileInfo = new FileInfo(Options.HistorySavePath);
                     _historyFileLastSavedSize = fileInfo.Length;
                 });
             }
+
+            static IEnumerable<string> ReadHistoryLinesImpl(string path, int historyCount)
+            {
+                const long offset_1mb = 1048576;
+                const long offset_05mb = 524288;
+
+                // 1mb content contains more than 34,000 history lines for a typical usage, which should be
+                // more than enough to cover 25,000 history records (a history record could be a multi-line
+                // command). Similarly, 0.5mb content should be enough to cover 10,000 history records.
+                // We optimize the file reading when the history count falls in those ranges. If the history
+                // count is even larger, which should be very rare, we just read all lines.
+                long offset = historyCount switch
+                {
+                    <= 10000 => offset_05mb,
+                    <= 25000 => offset_1mb,
+                    _ => 0,
+                };
+
+                using var fs = new FileStream(path, FileMode.Open);
+                using var sr = new StreamReader(fs);
+
+                if (offset > 0 && fs.Length > offset)
+                {
+                    // When the file size is larger than 1mb, we only read the last 1mb content from the end.
+                    int? b1 = null, b2 = null;
+                    fs.Seek(-offset, SeekOrigin.End);
+
+                    // After seeking, the current position may point at the middle of a history record, or even
+                    // a byte within a unicode char. So, we need to find the start of the next history record.
+                    while ((b2 = fs.ReadByte()) is not -1)
+                    {
+                        // Read bytes until we find the first newline ('\n' == 0xA) that is not right after a backtick ('`' == 0x60).
+                        // It means a separate full history record will start from the next byte.
+                        if (b2 is 0xA && b1.HasValue && b1 is not 0x60)
```

Comment:
Eh, it is not good practice to read unicode chars by bytes. Thus I could save the file in Utf16.

Since we want address an edge case - in most cases this file is in kilobytes or megabytes but not gigabytes - 
I suggest reading line by line, but if the file is really larger than 1.5GB ignore the first lines until the remainder of the file is around 500Mb.

Reply:
> it is not good practice to read unicode chars by bytes. Thus I could save the file in Utf16.

That's a good point. I will rework on this.

Reply:
> I suggest reading line by line, but if the file is really larger than 1.5GB ignore the first lines until the remainder of the file is around 500Mb.

I'd still like to only read partial content from the file, for 2 reasons:
1. reduce GC and unneeded operation as much as possible, also help get the prompt to user faster when the history file is large.
2. there is no good way to find out the reminder of the file when reading line by line, because the position of the underlying file stream is not going to point at the next byte to be read due to the block reading and caching of the stream reader.